examples/docker-rootful.yaml

```
@@ -44,16 +44,64 @@ provision:
     #!/bin/bash
     set -eux -o pipefail
     command -v docker >/dev/null 2>&1 && exit 0
-    if [ ! -e /etc/systemd/system/docker.socket.d/override.conf ]; then
-      mkdir -p /etc/systemd/system/docker.socket.d
+    readonly override_conf=/etc/systemd/system/docker.socket.d/override.conf
+    if [ ! -e "$override_conf" ]; then
+      mkdir -p $(dirname "$override_conf")
       # Alternatively we could just add the user to the "docker" group, but that requires restarting the user session
-      cat <<-EOF >/etc/systemd/system/docker.socket.d/override.conf
-      [Socket]
-      SocketUser={{.User}}
+      cat <<EOF >"$override_conf"
+    [Socket]
+    SocketUser={{.User}}
     EOF
     fi
     export DEBIAN_FRONTEND=noninteractive
     curl -fsSL https://get.docker.com | sh
+- mode: user  # configure docker under non-root user
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    command -v jq &>/dev/null || sudo apt-get install -y jq
+    readonly rootless_installed=$(systemctl --user list-unit-files docker.service &>/dev/null && echo true || echo false)
+
+    if [ "{{.Param.Rootful}}" = "true" ]; then
+      readonly config_dir="/etc/docker"
+      readonly systemctl="sudo systemctl"
+      readonly tee="sudo tee"
+
+      [ "$rootless_installed" != "true" ] || systemctl --user disable --now docker
```

To simplify the code I would especially try to avoid using boolean shortcut operator. Users unfamiliar with De Morgan's law find turning `if CONDITION; then ACTION; fi` into `! CONDITION || ACTION` unnatural, and it may be unobvious why they shouldn't write `CONDITION && ACTION` (it will cause the script to fail because of `set -o errexit`).

```suggestion
      if [ "$rootless_installed" == "true" ]; then
          systemctl --user disable --now docker
      fi
```

This feedback applies to other uses of `||` as well, but I won't repeat it on each instance.